src/pages/library/datasetBuilder/DatasetBuilder.ts

```diff
@@ -0,0 +1,419 @@
+import * as _ from 'lodash/fp';
+import React, { Fragment, useState } from 'react';
+import { div, h, h2, h3, label, li, ul } from 'react-hyperscript-helpers';
+import { ButtonPrimary, LabeledCheckbox, Link, spinnerOverlay } from 'src/components/common';
+import FooterWrapper from 'src/components/FooterWrapper';
+import { icon } from 'src/components/icons';
+import Modal from 'src/components/Modal';
+import TopBar from 'src/components/TopBar';
+import { DatasetBuilder, DatasetResponse } from 'src/libs/ajax/DatasetBuilder';
+import { useLoadedData } from 'src/libs/ajax/loaded-data/useLoadedData';
+import colors from 'src/libs/colors';
+import { useOnMount } from 'src/libs/react-utils';
+import * as Utils from 'src/libs/utils';
+import { StringInput } from 'src/pages/library/data-catalog/CreateDataset/CreateDatasetInputs';
+import {
+  PAGE_PADDING_HEIGHT,
+  PAGE_PADDING_WIDTH,
+  PREPACKAGED_CONCEPT_SETS,
+} from 'src/pages/library/datasetBuilder/constants';
+import { Cohort, ConceptSet, DatasetBuilderType } from 'src/pages/library/datasetBuilder/dataset-builder-types';
+import { DatasetBuilderHeader } from 'src/pages/library/datasetBuilder/DatasetBuilderHeader';
+import { datasetBuilderCohorts, datasetBuilderConceptSets } from 'src/pages/library/datasetBuilder/state';
+import { validate } from 'validate.js';
+
+const DatasetBuilderSelectorSubHeader = ({ children }) => div({ style: { fontSize: 12, fontWeight: 600 } }, children);
+
+interface DatasetBuilderObjectSet<T extends DatasetBuilderType> {
+  header: string;
+  values: T[];
+}
+
+interface DatasetBuilderSelectorProps<T extends DatasetBuilderType> {
+  number: number;
+  header: string;
+  subheader?: string;
+  datasetBuilderObjectSets: DatasetBuilderObjectSet<T>[];
+  selectedDatasetBuilderObjectSets: DatasetBuilderObjectSet<T>[];
+  onChange: (newDatasetBuilderObjectSets: DatasetBuilderObjectSet<T>[]) => void;
+  headerAction: any;
+  placeholder?: any;
+  style?: React.CSSProperties;
+}
+const DatasetBuilderSelector = <T extends DatasetBuilderType>({
+  number,
+  header,
+  subheader,
+  headerAction,
+  placeholder,
+  datasetBuilderObjectSets,
+  onChange,
+  selectedDatasetBuilderObjectSets,
+  style,
+}: DatasetBuilderSelectorProps<T>) => {
+  const isChecked = (datasetBuilderObjectSet, value) => {
+    return _.flow(
+      _.filter(
+        (selectedDatasetBuilderObjectSet: DatasetBuilderObjectSet<T>) =>
+          selectedDatasetBuilderObjectSet.header === datasetBuilderObjectSet.header
+      ),
+      _.flatMap((selectedDatasetBuilderObjectSet) => selectedDatasetBuilderObjectSet.values),
+      _.map((selectedValue) => selectedValue.name),
```

Comment:
I don't know if there's a good way to express this, but it seems wasteful that this is is only called in a context where we're calling it with the same `datasetBuilderObjectSet` every time in a loop, so the value computed by the `filter()`, `flatMap()` and `map()` is the same for every `value` in that set. One solution would be to find the matching selected object set and store the selected names in a `const` on line 112 below.